src/source/Ice/TurnConnection.c

```diff
@@ -781,34 +783,83 @@ STATUS turnConnectionDeliverChannelData(PTurnConnection pTurnConnection, PBYTE p
             channelMsgLen == paddedDataLen + TURN_DATA_CHANNEL_SEND_OVERHEAD, STATUS_INVALID_ARG);
     }
 
+    DLOGV("Data available from channel %u", channelNumber);
+    // Bail out early if we have no callback
+    CHK(pTurnConnection->turnConnectionCallbacks.applicationDataAvailableFn != NULL, retStatus);
+
     CHK_STATUS(doubleListGetHeadNode(pTurnConnection->turnPeerList, &pCurNode));
     while (pCurNode != NULL) {
         CHK_STATUS(doubleListGetNodeData(pCurNode, &data));
         pCurNode = pCurNode->pNext;
 
         pTurnPeer = (PTurnPeer) data;
         if (pTurnPeer->channelNumber == channelNumber) {
-            DLOGV("data available from channel %u", channelNumber);
-            if (pTurnConnection->turnConnectionCallbacks.applicationDataAvailableFn != NULL) {
-                pTurnConnection->turnConnectionCallbacks.applicationDataAvailableFn(
-                        pTurnConnection->turnConnectionCallbacks.customData,
-                        pTurnConnection->pControlChannel,
-                        pChannelMsg + TURN_DATA_CHANNEL_SEND_OVERHEAD,
-                        channelDataSize,
-                        &pTurnPeer->address,
-                        NULL);
-            }
-            break;
+            // Handling of the data is on a separate thread due to potential of a deadlock and in case the callback is
+            // not a prompt operation taking long period of time.
+            CHK(NULL != (pWrapper = (PConnectionDataAvailableWrapper)
+                    MEMALLOC(SIZEOF(ConnectionDataAvailableWrapper) + channelDataSize)), STATUS_NOT_ENOUGH_MEMORY);
+
+            // Store the data for the thread call
+            pWrapper->address = pTurnPeer->address;
+            pWrapper->customData = pTurnConnection->turnConnectionCallbacks.customData;
+            pWrapper->dataSize = channelDataSize;
+            pWrapper->channelNumber = channelNumber;
+            pWrapper->pTurnConnection = pTurnConnection;
+
+            // Set the pointer and store the data within the wrapper
+            pWrapper->pData = (PBYTE)(pWrapper + 1);
+            MEMCPY(pWrapper->pData, pChannelMsg + TURN_DATA_CHANNEL_SEND_OVERHEAD, channelDataSize);
+
+            CHK_STATUS(THREAD_CREATE(&callbackTid, turnDataAvailableWrapper, (PVOID) pWrapper));
+            CHK_STATUS(THREAD_DETACH(callbackTid));
+
+            // Stop the loop iteration
+            pCurNode = NULL;
         }
     }
 
 CleanUp:
 
     CHK_LOG_ERR_NV(retStatus);
 
+    if (STATUS_FAILED(retStatus) && pWrapper != NULL) {
```

Comment:
would it be memory leak if `retStatus` looks okay? as we won't `MEMFREE(pWrapper)`.

Reply:
No, in that case the ownership of the allocation is in the thread routine which would free it